pkg/util/version_history.go

```diff
@@ -0,0 +1,83 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-2020 Datadog, Inc.
+
+package util
+
+import (
+	"encoding/json"
+	"io/ioutil"
+	"path/filepath"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/config"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	"github.com/DataDog/datadog-agent/pkg/version"
+)
+
+type versionHistoryEntry struct {
+	Version   string    `json:"version"`
+	Timestamp time.Time `json:"timestamp"`
+}
+
+type versionHistoryEntries struct {
+	Entries []versionHistoryEntry `json:"entries"`
+}
+
+const maxVersionHistoryEntries = 60
+
+// LogVersionHistory loads version history file, append new entry if agent version is different than the last entry in the
+// JSON file, trim the file if too many entries then save the file.
+func LogVersionHistory() {
+	versionHistoryFilePath := filepath.Join(config.Datadog.GetString("logs_config.run_path"), "version-history.json")
+	logVersionHistoryToFile(versionHistoryFilePath, version.AgentVersion, time.Now().UTC())
+}
+
+func logVersionHistoryToFile(versionHistoryFilePath, agentVersion string, timestamp time.Time) {
+	file, err := ioutil.ReadFile(versionHistoryFilePath)
+
+	history := versionHistoryEntries{}
+
+	if err != nil {
+		log.Warnf("Cannot read file: %s, will create a new one. %v", versionHistoryFilePath, err)
+	} else {
+		err = json.Unmarshal(file, &history)
+		if err != nil {
+			// If file is in illegal format, ignore the error and regenerate the file.
+			log.Errorf("Cannot deserialize json file: %s. %v", versionHistoryFilePath, err)
+		}
+	}
+
+	newEntry := versionHistoryEntry{
+		Version:   agentVersion,
+		Timestamp: timestamp,
+	}
+
+	if len(history.Entries) == 0 || history.Entries[len(history.Entries)-1].Version != newEntry.Version {
+		// Only append the version info if no entry or this is different than the last entry.
+		history.Entries = append(history.Entries, newEntry)
+	} else {
+		// Otherwise no change is needed, just return.
+		return
+	}
+
+	// Trim entries if they grow beyond the max capacity.
```

Comment:
I would do a `log.Debug` here to have the trimming written down in the logs, might be helpful to the user in future diagnostics.

Reply:
I don't see why it would help. There's test code to ensure it work properly.